src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,130 @@
+(ns puppetlabs.puppetdb.threadpool
+  (:import [javax.jms ExceptionListener JMSException MessageListener Session
+            ConnectionFactory Connection Queue Message]
+           [java.util.concurrent Semaphore ThreadPoolExecutor TimeUnit SynchronousQueue
+            RejectedExecutionException ExecutorService]
+           [org.apache.commons.lang3.concurrent BasicThreadFactory BasicThreadFactory$Builder])
+  (:require [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n]
+            [slingshot.slingshot :refer [throw+]]))
+
+(def logging-exception-handler
+  "Exception handler that ensures any uncaught exception that occurs
+  on this thread is logged"
+  (reify Thread$UncaughtExceptionHandler
+    (uncaughtException [_ thread throwable]
+      (log/error throwable
+                 (i18n/trs "Error processing command")))))
```

Comment:
Maybe the thread id? Not really meaningful by itself, but you could differentiate errors between threads. We should never have this code get called, as we should be handling the exceptions properly in our code. This gets called if somehow an exception was to bubble up.